### PR TITLE
Add URI support to tarball path and xclbin path

### PR DESCRIPTION
**Pull request by sifive-benjamin-morse**
* This allows `config_hwdb.yaml` to use URI's for the `driver_tar` and `xclbin` keys.
* This also fixes an issue in PR #1299 regarding `server_hardware_config` being a string
* CI runs which previously used wget to download an `.xclbin` (such as `run-linux-poweroff-vitis.py`) can now directly specify the URL as a URI

#### UI / API Impact
Allows much easier collaboration with new developers as pre-compiled resources can be put on the web simplifying first time runs. 

#### Verilog / AGFI Compatibility
Should be no change.

### Contributor Checklist
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [ ] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous prints/debugging code?
- [ ] Did you state the UI / API impact?
- [ ] Did you specify the Verilog / AGFI compatibility impact?

- [ ] If applicable, did you regenerate and publicly share default AGFIs?

- [ ] If applicable, did you apply the `ci:fpga-deploy` label?

- [ ] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
Note: to run CI on PRs from forks, comment `@Mergifyio copy main` and manage the change from the new PR.
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [inline deploy/runtools/run_farm_deploy_managers.py:256]: I was hoping that when we added support for URIs that we would not download everything to the manager than then copy it to the runtime hosts, we would do the downloading in a distributed fashion on the runtime hosts.  When running large simulations of dozens of nodes, we're already constrained by the bandwidth available on the single manager host.

It also seems like this will change repeated `infrasetup` runs from being a checksum check done by the rsync to re-downloading the local copy and then doing the checksum to see that it doesn't need to be copied remotely.

If downloading locally makes it easier to add the URI support, we can distribute the downloading later to address the bottleneck.

File: deploy/runtools/runtime_config.py
Comment (timsnyder-siv) [inline deploy/runtools/runtime_config.py:120]: In https://github.com/firesim/firesim/pull/1140/files#r927042552 @davidbiancolin wanted to support things optionally being a filepath because URIs don't have a means for specifying a relative path.  His thinking may have changed on that though.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [inline deploy/runtools/run_farm_deploy_managers.py:147]: nit: if this needs to be a `@classmethod` it's customary to have the param be called `class` instead of `self` because that's what you'll be getting.

File: deploy/runtools/firesim_topology_elements.py
Comment (timsnyder-siv) [inline deploy/runtools/firesim_topology_elements.py:566]: Seems strange to me that these methods returning names relative to the `sim_slot_X` directory of a particular sim  are `@classmethod` of the `FireSimServerNode` class.

Comment (sifive-benjamin-morse) [reply deploy/runtools/firesim_topology_elements.py:566]: Ok, I wasn't exactly sure where to put these

File: conda-reqs.yaml
Comment (sifive-benjamin-morse) [reply conda-reqs.yaml:41]: conda-lock doesn't support updating one or two dependencies, without updating all others to the latest version. The solution of locking everything down came about after a discussion with @davidbiancolin : He said that the prebuilt drivers will break in future revisions if we have CPP dependencies constantly updating.

The reason I need to touch this at all, is so that I can add: 
    - fsspec, s3fs, aiohttp

Comment (timsnyder-siv) [reply conda-reqs.yaml:41]: the CPP dependencies are going to be bundled with the driver at build time in the tarball, aren't they?

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:256]: Benjamin pointed out my flawed understanding of `fabric.api.execute()`.  Distributing the downloads is out of scope for this PR.

File: conda-reqs.yaml
Comment (davidbiancolin) [reply conda-reqs.yaml:41]: I was worried about potential discrepancies between the two ways we build the driver. That may be ill-founded. We'll need to take this out-of-band.

File: .github/scripts/run-linux-poweroff-vitis.py
Comment (davidbiancolin) [inline .github/scripts/run-linux-poweroff-vitis.py:33]:
```suggestion
```

Nit: It's preferable to make this as a comment in the PR (to help with review) because the comment (without any accompanying code) will lose context over time.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (davidbiancolin) [inline deploy/runtools/run_farm_deploy_managers.py:783]: IIRC, I think you plan to remove the azure support before the final merge right?

File: deploy/runtools/runtime_config.py
Comment (davidbiancolin) [reply deploy/runtools/runtime_config.py:120]: I spoke to @sifive-benjamin-morse about this out-of-band (after you made this comment) and i really don't have a strong opinion anymore (so i'm deferring to Ben). I guess there's a strong possibility i'll want the feature when i start using the mangaer.

Whether non-URI, relative paths should be supported is probably a question best posed to the `firesim-dev` chat.

File: conda-reqs.yaml
Comment (timsnyder-siv) [reply conda-reqs.yaml:41]: conda-lock does support 'updating' something that is in the existing lockfile to a new version in a minimal fashion with it's `--update` flag.  However, it seems like you can't add new packages to the requirements spec and also use `--update` for them to minimally install new packages.  I am trying to get a discussion of this usecase going in https://github.com/conda/conda-lock/pull/355#discussion_r1111114862

I think I see how I could patch conda-lock to support using `--update` with new packages.  I'm not sure how long it would take me to work through actually making that enhancement given that I just started looking at the project's code.  However, I'm hoping that this is just a documentation issue with conda-lock and that it's already possible to do this or trivial for the original dev or maintainers.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (sifive-benjamin-morse) [reply deploy/runtools/run_farm_deploy_managers.py:783]: These are necessary but not sufficient changes to add Azure support. I don't have access to running an `.xclbin` flow anywhere except for Azure. So I was thinking of leaving these in for now.

**Comment by sifive-benjamin-morse** [comment]
I think that moving URI downloads out to a shared location is a bit complicated. Right now, with N run hosts, N URI downloads will occur. To fix this I would need to:
* edit `infrasetup_passes()` such that a URI download occurs outside of the loop.
* each node node can pick from shared or unique URI

My thoughts are:
* if the nodes are all the same type this is easy
* But there is no guarantee that each node is the same type making it complicated to share URIs.

Is there a simple solution here?


https://github.com/firesim/firesim/blob/6e993f5089a2290fba44ded39aadd972754c1083/deploy/runtools/firesim_topology_with_passes.py#L420-L436

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [inline deploy/runtools/run_farm_deploy_managers.py:837]: I can see how they were writing to a hardcoded tempfile on a possibly remote host and then opening it without copying it back.  In this new way of doing things, anything `xbutil` emits into `STDERR` could possibly cause difficult to understand and diagnose JSON parsing errors.

It better to pass `combine_stderr=False` (see [fabric docs](https://docs.fabfile.org/en/1.14/usage/interactivity.html#combine-streams))  to this invocation of `run()` and DEBUG log `xbutil_examine_json.stderr` if it isn't empty.

I don't remeember what uncaught `JSONDecodeError` stack traces look like but I vaguely remember them being unhelpful when the JSON being parsed is dynamically generated because they tell you where the error happened but don't print out what was parsed.  You might want to DEBUG log `xbutil_examine_json` or catch `JSONDecodeError` to raise our own meaningful `RuntimeError` message that includes the `JSONDecodeError.doc` in it's message

File: deploy/tests/test_utils.py
Comment (timsnyder-siv) [inline deploy/tests/test_utils.py:69]: I think you got this from the person who helped me write this test or me but it would be better to:
```suggestion
    test_dest_file_path.unlink()
```

File: deploy/tests/test_utils.py
Comment (timsnyder-siv) [inline deploy/tests/test_utils.py:54]:
```suggestion
    assert test_dest_file_path.exists(), f"{test_dest_file_path} was not created."
```

File: deploy/tests/test_utils.py
Comment (timsnyder-siv) [inline deploy/tests/test_utils.py:47]:
```suggestion
    test_dest_file_path.unlink(missing_ok=True)
```

File: deploy/tests/test_utils.py
Comment (timsnyder-siv) [inline deploy/tests/test_utils.py:5]:
```suggestion
```

I think the only uses of `os` were misguided uses of stuff that are better done with the `Path` methods.

File: docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst
Comment (timsnyder-siv) [inline docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst:549]: Now that we have multiple things that can be a URI, it might be better to move this to a section at the top of this page or it's own page and link to it from anything that can use a URI.

File: scripts/machine-launch-script.sh
Comment (timsnyder-siv) [inline scripts/machine-launch-script.sh:235]: The requirements in this file a minimal and only serve to give the user enough of an environment to be able to clone the firesim repo and run `build_setup.sh`.  I see that other stuff has snuck in here since I last looked but are `fsspec` and `s3fs` needed before we have the manager?

Comment (timsnyder-siv) [reply scripts/machine-launch-script.sh:235]: I think they may have been here in the original implementation of this PR because we hadn't yet moved the full requirements into conda-reqs.yml.

**Review by timsnyder-siv — COMMENTED** [review]
Thanks for pushing this toward the finish line.  I only saw a couple of things I had questions or concerns about.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (russell-horvath) [inline deploy/runtools/run_farm_deploy_managers.py:124]: Is this comment duplicated? Its basically the same as the one inside the function

**Review by russell-horvath — COMMENTED** [review]
small thing in comment above*

File: deploy/runtools/run_farm_deploy_managers.py
Comment (sifive-benjamin-morse) [reply deploy/runtools/run_farm_deploy_managers.py:327]: No. Each InstanceDeployManager object maintains a member, uri_list, which is a list of URI that need to be copied. This is not shared. Neighbor URI's are not copied over. Is there something I can change to make this more clear?

Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:325]: I think I was confused by the singular `get_local_uri_path`.  If it was named `get_local_uri_paths`, it might have helped.  I just wanted to make sure you weren't copying the whole cache by passing rsync a single directory name to copy.

File: scripts/machine-launch-script.sh
Comment (sifive-benjamin-morse) [reply scripts/machine-launch-script.sh:235]: Without these changes, `setup-manager` fails with this import path.
```
/opt/conda/envs/firesim/lib/python3.11/site-packages/paramiko/transport.py:178: CryptographyDeprecationWarning: Blowfish has been deprecated
  'class': algorithms.Blowfish,
Traceback (most recent call last):
  File "/home/centos/actions-runner-0/_work/firesim/firesim/./.github/scripts/install-firesim-pem.py", line 7, in <module>
    from common import manager_home_dir, manager_fsim_pem, set_fabric_firesim_pem
  File "/home/centos/actions-runner-0/_work/firesim/firesim/.github/scripts/common.py", line 6, in <module>
    from platform_lib import PlatformLib, AWSPlatformLib, AzurePlatformLib, Platform
  File "/home/centos/actions-runner-0/_work/firesim/firesim/.github/scripts/platform_lib.py", line 24, in <module>
    from awstools.awstools import get_instances_with_filter
  File "/home/centos/actions-runner-0/_work/firesim/firesim/deploy/awstools/awstools.py", line [33](https://github.com/firesim/firesim/actions/runs/4296629292/jobs/7488725823#step:3:34), in <module>
    from util.io import firesim_input
  File "/home/centos/actions-runner-0/_work/firesim/firesim/deploy/util/io.py", line 3, in <module>
    from fsspec.core import url_to_fs, open_local # type: ignore
    ```

(from https://github.com/firesim/firesim/actions/runs/4296629292/jobs/7488725823)

File: deploy/runtools/run_farm_deploy_managers.py
Comment (sifive-benjamin-morse) [reply deploy/runtools/run_farm_deploy_managers.py:605]: @davidbiancolin do you have a better place to put these?

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 4296740217. Verify CI state before submitting PR.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (sifive-benjamin-morse) [reply deploy/runtools/run_farm_deploy_managers.py:65]: OK this makes sense. I will try to paint a better picture for reviewers in the PR comments

File: deploy/runtools/run_farm_deploy_managers.py
Comment (sifive-benjamin-morse) [reply deploy/runtools/run_farm_deploy_managers.py]: I got this code by grabbing lots of things from https://github.com/firesim/firesim/compare/main...tsondhi/azure_validation_backup

I believe that this change was combined with a change to add a new property `sim_command_requires_sudo()` I

File: deploy/runtools/run_farm_deploy_managers.py
Comment (davidbiancolin) [reply deploy/runtools/run_farm_deploy_managers.py:607]: As far as location goes, i think it's OK to keep them in the class, either as methods or whatever the pythonic way is of definine class-scoped constants, though the XCLBIN method should only be defined in the VitisDeployManager or w.e if possible. 

I think the naming of the methods is more critical: `get_slot_relative_<name>` might be more along the lines of what @timsnyder-siv is asking for

File: scripts/machine-launch-script.sh
Comment (timsnyder-siv) [reply scripts/machine-launch-script.sh:235]: Thanks.  I think I got Abe's attention to try and dedup the requirements a bit in #1445

**Review by timsnyder-siv — APPROVED** [review]
Thanks again for getting this really implemented @sifive-benjamin-morse

File: deploy/util/io.py
Comment (timsnyder-siv) [inline deploy/util/io.py:54]: Are you sure you want the `url_to_fs()` call in the try block?  It seems like exceptions arising here may be indicative of a larger issue that shouldn't be retried?

Comment (sifive-benjamin-morse) [reply deploy/util/io.py:54]: For background, these two runs: 
* https://github.com/firesim/firesim/actions/runs/4301469200/jobs/7498730505
* https://github.com/firesim/firesim/actions/runs/4300595364/jobs/7496935845

fail for two different reasons on the same line:
* `aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host firesim-ci-vitis-xclbins.s3.***.amazonaws.com:443 ssl:default [Name or service not known]`
* `aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host firesim-ci-vitis-xclbins.s3.***.amazonaws.com:443 ssl:default [Temporary failure in name resolution]`

I added url_to_fs() inside the block because I wasn't sure if it did the DNS resolution or not. I will move it out. Thanks for reviewing this

**Pull request merged**
